### PR TITLE
Require go 1.17+

### PR DESCRIPTION
**Pull request by BenTheElder**
This enables better dependency management via https://go.dev/ref/mod#graph-pruning

We previously required 1.16+

We don't want to require especially recent versions because that will break users with `go install` using older packaged golang.

We use currently supported much more recent go for our own builds.

NOTE: there are no docs changes yet, because the docs have users installing v0.20 which is still compatible with go 1.16

### CODE REVIEW

**Comment by BenTheElder** [comment]
had to update the tools while working on this, looks like we picked up some new linter issues, working on resolving.

File: images/kindnetd/cmd/kindnetd/main.go [hunk 1]
```diff
@@ -252,7 +252,7 @@ func makeNodesReconciler(cniConfig *CNIConfigWriter, hostIP string, ipFamily IPF
 
 		// obtain the PodCIDR gateway
 		var nodeIPv4, nodeIPv6 string
-		for _, ip := range nodeIPs.List() {
+		for _, ip := range sets.List(nodeIPs) {
```
Comment (BenTheElder) [inline images/kindnetd/cmd/kindnetd/main.go:255]: [this is the non-deprecated way to use the sets package]

File: go.mod [hunk 2]
```diff
@@ -1,6 +1,14 @@
...
-go 1.16
+// NOTE: This is the go language version, NOT the compiler version.
+//
+// This controls the *minimum* required go version and therefore available Go
+// language features.
+//
+// See ./.go-version for the go compiler version used when building binaries
+//
+// https://go.dev/doc/modules/gomod-ref#go
+go 1.17
```
Comment (aojea) [inline go.mod:11]: well , this is changing in 1.21

> For example, a go.mod that says go 1.21.0 with no toolchain line is interpreted as if it had a toolchain go1.21.0 line.

https://go.dev/doc/toolchain

**Comment by aojea** [comment]
just the question about the behavior change in golang with 1.21, the rest LGTM

File: go.mod (hunk 2 above)
Comment (BenTheElder) [reply go.mod:11]: We should rework everything around .go-version to switch to this toolchain line.

Comment (BenTheElder) [reply go.mod:11]: Thanks, I forgot about this.

**Comment by BenTheElder** [comment]
I'm going to finish this up today, ahead of generally bumping things for a v0.21 cut.

**Comment by BenTheElder** [comment]
It's going to be a bit awkward to re-work this around plumbing GOTOOLCHAIN and I'm not sure what the best answer is, we've definitely got downstreams depending on patching .go-version. It might be best to just break that in favor of `toolchain` in the modules and put a release note. Need to think a bit.

**Comment by BenTheElder** [comment]
I'm punting this to the next release because we really need to get that runc CVE out, but some veersion of the changes here probably need to be high on the list longterm as part of switching to go1.21+

**Comment by BenTheElder** [comment]
Actually, I think the most reasonable answer is just to set both .go-version and toolchain to match (after confirming older go will ignore toolchain in the module file) and then deprecated .go-version and teach the gimme fallback to read it (though just getting 1.21 would be good enough since 1.21 will turn around and download toolchain).

**Comment by BenTheElder** [comment]
rebased and started forcing GOTOOLCHAIN="go${GOVERSION}" everywhere for now.

File: go.mod (hunk 2 above)
Comment (BenTheElder) [reply go.mod:11]: so, I added a commit to deal with this, but the remaining downside (Also true in k/k currently), is that IDEs won't see all our bash hacks.

what we probably need to do is pivot to setting toolchain in the go mods and then also for the image builds of external binaries, but that's going to force us to decentralize setting the go version unfortunately.

Comment (BenTheElder) [reply go.mod:11]: the current PR is enough to ensure our builds / tools work at least

**Comment by BenTheElder** [comment]
> pull-kind-conformance-parallel-ga-only — Pod got deleted unexpectedly   
/retest

**Comment by BenTheElder** [comment]
@aojea go 1.22 released so go 1.20 will be EOL and we'll have to sort this out. Will aim to look again later this week for better iteration, but the current version should be usable.

File: .go-version [hunk 3]
```diff
@@ -1 +1 @@
-1.20.13
+1.21.6
```
Comment (aojea) [inline .go-version:1]: this should be 1.22 no?

Comment (BenTheElder) [reply .go-version:1]: this PR has been around awaiting review for a while :-)

Comment (BenTheElder) [reply .go-version:1]: we're still on 1.20.13 @ HEAD

File: images/base/Dockerfile [hunk 4]
```diff
@@ -141,6 +143,7 @@ RUN git clone --filter=tree:0 "${RUNC_CLONE_URL}" /runc \
     && cd /runc \
     && git checkout "${RUNC_VERSION}" \
     && eval "$(gimme "${GO_VERSION}")" \
+    && export GOTOOLCHAIN="go${GO_VERSION}" \
```
Comment (aojea) [inline images/base/Dockerfile:146]: line159 seems is missing this export too

File: images/kindnetd/go.mod [hunk 5]
```diff
@@ -1,6 +1,6 @@
 module sigs.k8s.io/kind/images/kindnetd
 
-go 1.13
+go 1.18
```
Comment (aojea) [inline images/kindnetd/go.mod:3]: why don't you move to latest here then?

Comment (BenTheElder) [reply images/kindnetd/go.mod:3]: it matched kubernetes at the time. note that this is not the go version used to compile, this is the language level features used.

File: images/base/Dockerfile (hunk 4 above)
Comment (BenTheElder) [reply images/base/Dockerfile:146]: yes. I will add this and bump up the go versions again

File: images/kindnetd/go.mod (hunk 5 above)
Comment (BenTheElder) [reply images/kindnetd/go.mod:3]: let's leave this for a later PR, this PR is already a lot and I don't want to deal with any breakage while trying to get the rest of the PR in.

File: images/base/Dockerfile (hunk 4 above)
Comment (BenTheElder) [reply images/base/Dockerfile:149]: fixed this one.

**Comment by BenTheElder** [comment]
TODO: 
- rebump everything to latest
- explore using go.mod gotoolchain instead (need to handle things like the gimme fallback, and approach to keep these in sync or not)

leaving for follow-ups, I want to get in the core "we're unblocked to manage go versions with 1.21+ without switching language version to 1.21+" so we can collectively iterate from there.

**Comment by BenTheElder** [comment]
There is a new linter failure from nerdctl code, fixing ...

**Comment by BenTheElder** [comment]
/retest
https://prow.k8s.io/view/gs/kubernetes-jenkins/pr-logs/pull/kubernetes-sigs_kind/3335/pull-kind-e2e-kubernetes-1-28/1774934849065324544

network glitch during cloning, not related.

**Comment by BenTheElder** [comment]
hrm, that's not good.

https://prow.k8s.io/view/gs/kubernetes-jenkins/logs/post-kind-push-base-image/1774949333360185344

**Comment by BenTheElder** [comment]
ah, I know what this is. patch coming in a bit (dealing with some internal work atm)

**Comment by BenTheElder** [comment]
base is building again

**Pull request merged**